src/rgw/rgw_cache.cc

```diff
@@ -344,3 +359,424 @@ ObjectCache::~ObjectCache()
+  remove(location.c_str());
+  return freed_size;
+}
+
+size_t DataCache::lru_eviction(){
+
+  int n_entries = 0;
+  int random_index = 0;
+  size_t freed_size = 0;
+  ChunkDataInfo *del_entry;
+  string del_oid, location;
+
+  eviction_lock.Lock();
+  del_entry = tail;
+  lru_remove(del_entry);
+  eviction_lock.Unlock();
+
+  cache_lock.Lock();
+  n_entries = cache_map.size();
+  if (n_entries <= 0){
+    cache_lock.Unlock();
+    return -1;
+    }
+  del_oid = del_entry->oid;
+  ldout(cct, 20) << "Engage1: lru_eviction: oid to remove" << del_oid << dendl;
+  map<string, ChunkDataInfo*>::iterator iter = cache_map.find(del_entry->oid);
+  if (iter != cache_map.end()) {
+    cache_map.erase(del_oid); // oid
+  }
+  cache_lock.Unlock();
+  freed_size = del_entry->size;
+  free(del_entry);
+  location = cct->_conf->rgw_datacache_persistent_path + del_oid; /*replace tmp with the correct path from config file*/
+  remove(location.c_str());
+  return freed_size;
+}
+
+
+void DataCache::remote_io(struct librados::L2CacheRequest *l2request ) {
+  ldout(cct, 20) << "Engage1: Add task to remote IO" << dendl;
+  tp->addTask(new HttpL2Request(l2request, cct));
+}
+
+
+std::vector<string> split(const std::string &s, char * delim) {
+  stringstream ss(s);
+  std::string item;
+  std::vector<string> tokens;
+  while (getline(ss, item, (*delim))) {
+    tokens.push_back(item);
+  }
+  return tokens;
+}
+
+void DataCache::push_l2_request(struct librados::L2CacheRequest *l2request ) {
+  tp->addTask(new HttpL2Request(l2request, cct));
+}
+
+static size_t _l2_response_cb(void *ptr, size_t size, size_t nmemb, void* param) {
+  librados::L2CacheRequest *req = static_cast<librados::L2CacheRequest *>(param);
+  req->pbl->append((char *)ptr, size*nmemb);
+  return size*nmemb;
+}
+
+void HttpL2Request::run() {
+
+  get_obj_data *d = (get_obj_data *)req->op_data;
+  int n_retries =  cct->_conf->rgw_l2_request_thread_num;
+  int r = 0;
+  
+  for (int i=0; i<n_retries; i++ ){
+    if(!(r = submit_http_request())){
+      d->cache_aio_completion_cb(req);
+      return;
+    }
+    if (r == ECANCELED) {
+      return;
+    }
+
+  }
+}
+
+int HttpL2Request::submit_http_request () {
+
+  CURLcode res;
+  string auth_token;
+  string range = std::to_string(req->ofs + req->read_ofs)+ "-"+ std::to_string(req->ofs + req->read_ofs + req->len - 1);
+  struct curl_slist *header = NULL;
+  get_obj_data *d = (get_obj_data *)req->op_data;
+  struct req_state *s;
+  ((RGWGetObj_CB *)(d->client_cb))->get_req_info(req->dest, s->info.request_uri, auth_token);
+
+  if (s->dialect == "s3") {
+    RGWEnv env;
+    req_info info(cct, &env);
+    memcpy(&info, &s->info, sizeof(info));
+    memcpy(&env, s->info.env, sizeof(env));
+    info.env = &env;
+    std::string access_key_id;
+    if (true) { //(!s->http_auth || !(*s->http_auth)) { FIXME: The S3 Authentication has been changed! FIX IT
```

Comment:
so, did these get fixed later?

Reply:
This is part of the L2 code, have not tested it yet, most probably will need to be fixed.